internal/builder/builder.go

```diff
@@ -178,30 +179,47 @@ func (b *Builder) Build(ctx context.Context) error {
 
 	processed := map[string]struct{}{}
 	rootMap := map[string]*Source{}
+	record := func(i int, name string, fs_ fs.FS) {
+		prefix := name
+		if prefix == "" || i > 0 {
+			prefix += strconv.Itoa(i)
+		}
+		prefix = util.Escape(prefix)
+		buildSources = append(buildSources, build{prefix: prefix, fsys: fs_})
+	}
 
 	for len(toProcess) > 0 {
 		var next *Source
 		next, toProcess = toProcess[0], toProcess[1:]
 		var newRoots []ast.Ref
 
+		checkRoots := func(fs_ fs.FS) error {
+			rs, err := getRegoAndJSONRoots(fs_)
+			if err != nil {
+				return fmt.Errorf("%v: %w", next.Name, err)
+			}
+			newRoots = append(newRoots, rs...)
+			return nil
+		}
+
+		for i, fs0 := range next.FSes {
+			if err := checkRoots(fs0); err != nil {
```

Comment:
Do we need to check for `b.excluded` here as well? I guess for the purpose of `builtins` probably not, but if it is ever reused for some other purpose we might run into issues.

Could make sense to implement a `source.Filesystems` method that unifies `source.FSes` and `source.Dirs` and does the source-level filtering. Then the filtering for `b.excluded` could be a FilterFS on top of that.

Reply:
Heh that's a nice refactoring idea. I'll give it a try!